src/unread/unreadHuddlesReducer.js

```diff
@@ -24,7 +24,7 @@ const eventNewMessage = (state, action) => {
     return state;
   }
 
-  if (action.ownUserId === action.message.sender_id) {
+  if (action.message.flags?.includes('read')) {
```

Comment:
Looking at the comment on `.flags` in `Message` (which it gets from `BaseMessage` now), I see that we can expect `flags` to always be present here:

```js
  /**
   * The `flags` story is a bit complicated:
   *  * Absent in `event.message` for a `message` event... but instead
   *    (confusingly) there is an `event.flags`.
   *  * Present under `EVENT_NEW_MESSAGE`.
   *  * Present under `MESSAGE_FETCH_COMPLETE`, and in the server `/message`
   *    response that action is based on.
   *  * Absent in the Redux `state.messages`; we move the information to a
   *    separate subtree `state.flags`.
   */
```

Probably a check like the one in `unreadMentionsReducer` would be cleaner (fewer cases to reason about) than accepting a missing `flags` as valid input:

```js
      if (!flags) {
        throw new Error('action.message.flags should be defined.');
      }
```

though I would do it more concisely with [`invariant`](https://github.com/zulip/zulip-mobile/blob/master/docs/style.md#invariant-assertions), which I think we started using regularly after that bit of code was written.

Reply:
I took a look at how to do this, but I'm actually even more confused now: I don't understand how `flags` ends up being a property of the message, rather than the event. From this, it sounds like the server will send down a `EVENT_NEW_MESSAGE` with `flags` set on the event, but the code that I wrote looks for `flags` under the `message` in the event, and I don't understand why that works.

Specifically, the blocker to making this change is that when I try to make it, a bunch of tests fail, since the tests don't set the `flags` on the message. Do you know what's going on with that?

Reply:
The thing that fills in `action.message.flags` in live, non-test code is the commented line in this bit of src/events/eventToAction.js:

```js
    case 'message':
      return {
        type: EVENT_NEW_MESSAGE,
        id: event.id,
        message: {
          ...event.message,
          // Move `flags` key from `event` to `event.message` for
          // consistency; default to empty if `event.flags` is not set.
          flags: event.message.flags ?? event.flags ?? [],
          avatar_url: AvatarURL.fromUserOrBotData({
            rawAvatarUrl: event.message.avatar_url,
            email: event.message.sender_email,
            userId: event.message.sender_id,
            realm: getCurrentRealm(state),
          }),
        },
        local_message_id: event.local_message_id,
        caughtUp: state.caughtUp,
        ownUserId: getOwnUserId(state),
      };
```

In test code, it looks like `eg.streamMessage` and `eg.pmMessage` don't put `.flags` in the `Message` objects they return.

:(

There's this line in `messagePropertiesBase` in exampleData.js (that's a thing those functions use to fill in boring properties common to all example `Message` objects):

```js
// flags omitted
```

And `eg.streamMessage` and `eg.pmMessage` don't bother to fill it in either.

We could have `eg.pmMessage` and `eg.streamMessage` take a `flags` param; that seems useful here and elsewhere. I'm not sure if the right fallback for when `flags` isn't passed is to omit `flags` or use an empty array, though. I'm not sure if there are current tests that depend on `flags` being omitted. If there are, they probably deal with this case:

```
   *  * Absent in the Redux `state.messages`; we move the information to a
   *    separate subtree `state.flags`.
```

(That's from the comment on `Message.flags` in modelTypes.js.)

Hmm, and after having `eg.pmMessage` and `eg.streamMessage` return an (empty-array) `.flags`, I see some unhelpful-looking Jest messages. I expect those'll be made actionable when we start using Jest 26, in #4700, and we get facebook/jest#10414. (I recall that we specifically wanted that in 64eaab4617af09e82d9a87d76edec84257d386b2, when we "took" Jest 26, but then we silently slipped back to 25 a few weeks later, in c4fca9d9dcdab1fab0b4674b673cb8ee22eeb0dc, when we added `jest-expo` back; `jest-expo`'s effect there is described in #4700.)

Reply:
Unless you'd like to push further, I'd say it's fine to leave a quick comment in each place, saying that we're prepared to error on a missing `action.message.flags` in `EVENT_NEW_MESSAGE`, but our tests will have to catch up before we can do that.

Reply:
Yeah, the situation where there are subtly-different types of message objects in different contexts (with or without `flags`) is kind of confusing.

I think omitting `flags` is the right thing for `eg.pmMessage` and `eg.streamMessage`, because that's what's expected in places like `state.messages`. But we should certainly make it easier for tests to get data that supplies it when that's what's needed.

I've just sent #4760 which generalizes `mkMessageAction` -- a test helper whose basically one job is to supply a default `flags` of `[]` -- into a new `eg.mkActionEventNewMessage`. Then after converting to that everywhere, we get to turn these into assertions.

Reply:
> Hmm, and after having `eg.pmMessage` and `eg.streamMessage` return an (empty-array) `.flags`, I see some unhelpful-looking Jest messages. I expect those'll be made actionable when we start using Jest 26, in #4700, and we get [facebook/jest#10414](https://github.com/facebook/jest/pull/10414).

FTR I just tried this, now that #4700 is in. The Jest error messages are clear now, and all center on diffs like this:
```diff
    -   "flags": Array [],
```
They're all in `messagesReducer-test`, and have the form "this message gotten from the resulting messages state didn't have a `flags`, and one was expected". Having no `flags` is the correct behavior there, so this illustrates that we need to continue to have an easy way for test code to make a `Message` object with no `flags`.